### ISSUE TITLE
ahci_drv: devices not reported

### ISSUE THREAD

**Issue by cproc** [genodelabs/genode#5088]
When I tried to run Sculpt on a PC with an ASUS Q170T main board, the AHCI driver did not report the connected SATA SSDs, but showed them as `off` in the log:

```
[init -> drivers -> dynamic -> ahci_drv] --- Starting AHCI driver ---
[init -> drivers -> dynamic -> ahci_drv] version: major=0x1 minor=0x301
[init -> drivers -> dynamic -> ahci_drv] command slots: 32
[init -> drivers -> dynamic -> ahci_drv] native command queuing: yes
[init -> drivers -> dynamic -> ahci_drv] 64-bit support: yes
[init -> drivers -> dynamic -> ahci_drv] number of ports: 6
[init -> drivers -> dynamic -> ahci_drv] 		#0: off (ATA)
[init -> drivers -> dynamic -> ahci_drv] 		#1:  off (unknown device signature)
[init -> drivers -> dynamic -> ahci_drv] 		#2:  off (unknown device signature)
[init -> drivers -> dynamic -> ahci_drv] 		#3: off (ATA)
[init -> drivers -> dynamic -> ahci_drv] 		#4:  off (unknown device signature)
[init -> drivers -> dynamic -> ahci_drv] 		#5:  off (unknown device signature)
```

The error symptom in the driver is that setting the `Cmd::Fre` bit to 0 does not cause the `Cmd::Fr` bit to change to 0 as well:

https://github.com/genodelabs/genode/blob/c4679e7af644f2ff8d07b9b48bc2c79eb57a79cc/repos/os/src/drivers/ahci/ahci.h#L880-L885

The problem appears to be related to the port reset at the beginning of the `Port()` constructor:

https://github.com/genodelabs/genode/blob/c4679e7af644f2ff8d07b9b48bc2c79eb57a79cc/repos/os/src/drivers/ahci/ahci.h#L549-L555

When I tried the `Cmd::Fre` bit change before this `reset()` call, the `Cmd::Fr` bit changed to 0 as expected.

The AHCI specification mentions that "software should write all 1s to the PxSERR register to clear any bits that were set as part of the port reset.", which is done after the second port reset in the `init()` function, but not after the first port reset in the `Port()` constructor:

https://github.com/genodelabs/genode/blob/c4679e7af644f2ff8d07b9b48bc2c79eb57a79cc/repos/os/src/drivers/ahci/ahci.h#L856-L860

When I added a `clear_serr()` call in the `Port()` constructor, the `Cmd::Fr` bit change was successful, but an IOMMU error message appeared during `clear_serr()`:

```
[ 0] IOMMU:0xffffffff817ea0b0 FRR:0 FR:0x5 BDF:0:17:0 FI:0x0009d000 (0)
```

This error message does not occur during the second `clear_serr()` call in the `init()` function and apparently got "fixed" by the `setup_memory()` call before the second port reset.

Removing the first `reset()` call from the `Port()` constructor appears to solve the problem without an IOMMU error, but I'm not sure if it is the correct fix.

**Comment by cproc** [comment]
> Have the changes in #5101 positive impact on this issue?

No, the problem still exists. The problematic first port reset has now moved to `Port::reinit()`.

**Comment by chelmuth** [comment]
[This discussion](https://genode.discourse.group/t/ahci-driver-partition-setup-in-qemu-hardware/97/8) in the user forum hints we should address this issue...

**Comment by G0retZ** [comment]
Any plans on fixing this?

**Comment by G0retZ** [comment]
@nfeske I created a PR that should fix this issue ✌️

**Comment by ssumpf** [comment]
@G0retZ: Could you give [7bb7321](https://github.com/ssumpf/genode/commit/7bb73219e9d1a0886c2c22f01d5b67bc03b7c121) a try and see if it changes anything? It is based on current master and can be found at my  issue5088_ahci_reset branch.

**Comment by G0retZ** [comment]
@ssumpf let me try that 👍 

On my question about any cons of this (spec's compliance or security) AI flagged only this 🙂 

> The practical downside is the extra COMRESET cost on every boot and S3 resume — a performance and log-noise trade-off, not a correctness or security one.

**Comment by G0retZ** [comment]
@ssumpf I tried your commit, it also works for me on boot 👍

The only thing that is common for all sculpt builds from `master` for me is that "Hard reboot" doesn't work. It just hangs the system. Same thing with "Hard power down" and "Suspend" after I enable acpi launcher.
Probably it's because I didn't try to rebuild all the packages into my own repo and as [mentioned here](https://genode.discourse.group/t/26-02-can-t-install-anything/304/6) this is an experimental combination for Sculpt...

But once again this hanging issue is common for the current `master` without any changes, with my commit and with yours.
Means it's not caused by our changes in ahci driver.

Other than that your commit works good on boot! 👍

**Comment by ssumpf** [comment]
@cproc: Could you give [4b5aaa0](https://github.com/ssumpf/genode/commit/4b5aaa00ca6c07ee6d94c36b3c2bd07c3d699714) on your ASUS Q170T a try?

**Comment by G0retZ** [comment]
@ssumpf I did additional tests today morning.
I built from `sculpt-25.10.1` tag and both fixes (from my PR and from your branch) worked for me on it also.

So both your and mine changes work for my setup on both `sculpt-25.10.1` tag and `master` branch. 👍 

I don't remember if I mentioned, I run on `hw` kernel, not on `nova`. (Though my disks are not visible on current nova builds also)

There's some issue with power down and reset features, but those are unrelated to these changes because they are there even without any changes...

**Comment by G0retZ** [comment]
> [@cproc](https://github.com/cproc): Could you give [4b5aaa0](https://github.com/ssumpf/genode/commit/4b5aaa00ca6c07ee6d94c36b3c2bd07c3d699714) on your ASUS Q170T a try?

Is worth to try this on my end also and report the results?

**Comment by ssumpf** [comment]
> > [@cproc](https://github.com/cproc): Could you give [4b5aaa0](https://github.com/ssumpf/genode/commit/4b5aaa00ca6c07ee6d94c36b3c2bd07c3d699714) on your ASUS Q170T a try?
> 
> Is worth to try this on my end also and report the results?

Sure, but it is based on current staging. You can try to cherry-pick the commit on master or Sculpt-25.10.

**Comment by G0retZ** [comment]
> > > [@cproc](https://github.com/cproc): Could you give [4b5aaa0](https://github.com/ssumpf/genode/commit/4b5aaa00ca6c07ee6d94c36b3c2bd07c3d699714) on your ASUS Q170T a try?
> > 
> > 
> > Is worth to try this on my end also and report the results?
> 
> Sure, but it is based on current staging. You can try to cherry-pick the commit on master or Sculpt-25.10.

Just tried this change on `sculpt-25.10.1` tag - worked fine for me 👍

**Comment by ssumpf** [comment]
> Just tried this change on `sculpt-25.10.1` tag - worked fine for me 👍

Great. We will wait for @cproc's results and in case it works for him as well this can enter staging from my side.

**Comment by cproc** [comment]
> [@cproc](https://github.com/cproc): Could you give [4b5aaa0](https://github.com/ssumpf/genode/commit/4b5aaa00ca6c07ee6d94c36b3c2bd07c3d699714) on your ASUS Q170T a try?

works here too 👍

**Comment by chelmuth** [comment]
Merged c8647afc0278465c8016fbc30cb73678db579b30 to staging just now.

**Comment by ssumpf** [comment]
@cproc, @G0retZ: Thanks for your insights and for testing.

**Comment by ssumpf** [comment]
@G0retZ: The power down/reset problems could be related to ACPI. This should be addressed by a separate issue.

**Comment by nfeske** [comment]
Fixed in master.
